File: AulaTPII/F217/GoF/Estruturais/Decorator/Decorator01.py
# Componente 
class ComponentNotificador:
    def enviar(self, mensagem):
        print(f"Enviando Mensagem: {mensagem}")

# Decorator
class DecoratorNotificador:
    def __init__(self, componente):
        self.componente = componente

    def enviar(self, mensagem):
        self.componente.enviar(mensagem)

# Decorator Concreta 1
class DecoratorNotificadorSMS(DecoratorNotificador):
    def enviar(self, mensagem):
        super().enviar(mensagem)
        print(f"Enviar mensagem SMS: {mensagem}")

# Decorator Concreta 2
class DecoratorNotificadorFaceBook(DecoratorNotificador):
    def enviar(self, mensagem):
        super().enviar(mensagem)
        print(f"Enviar mensagem FaceBook: {mensagem}")

# Cliente #################################
notificador = ComponentNotificador()
notificadorSMS = DecoratorNotificadorSMS(notificador)
notificadorFaceBook = DecoratorNotificadorFaceBook(notificador)

notificadorFaceBook.enviar("Bom dia")

File: AulaTPII/F217/GoF/Estruturais/Decorator/Decorator01.js
 //componente
class ComponetNotificador{
    enviar(mensagem){
        console.log(`Enviando Mensagem: ${mensagem}`)
    }
}
 
//Decorator Base
class DecoratorNotificador{
    constructor(componente){
        this.componente = componente
    }
    enviar(mensagem){
        this.componente.enviar(mensagem);
    }
}
 
//Decorar Concreta 1
class DecoratorNotificadorSMS extends DecoratorNotificador{
    enviar(mensagem){
        super.enviar(mensagem);
        console.log(`Enviar mensagem SMS:  ${mensagem}`)
    }
}
//Decorar Concreta 2
class DecoratorNotificadorFaceBook extends DecoratorNotificador{
    enviar(mensagem){
        super.enviar(mensagem);
        console.log(`Enviar mensagem FaceBook:  ${mensagem}`)
    }
}
 
//Cliente
const notificador = new ComponetNotificador()
const notificadorSMS = new DecoratorNotificadorSMS(notificador)
const notificadoFaceBook = new DecoratorNotificadorFaceBook(notificador)
 
notificadoFaceBook.enviar("Bom Dia")